lib/std/fs/path.zig

```diff
@@ -1318,3 +1333,552 @@ test "stem" {
     try testStem(" ", " ");
     try testStem("", "");
 }
+
+/// A path component iterator that can move forwards and backwards.
+/// The 'root' of the path (`/` for POSIX, things like `C:\`, `\\server\share\`, etc
+/// for Windows) is treated specially and will not be iterated over when iterating backwards.
+/// Multiple sequential path separators are skipped (treated as a single separtor)
+/// when iterating.
+/// All returned component names/paths are slices of the original path--there is no
+/// normalization of paths performed while iterating.
+pub fn ComponentIterator(comptime path_type: PathType, comptime T: type) type {
+    return struct {
+        path: []const T,
+        root_end_index: usize = 0,
+        start_index: usize = 0,
+        end_index: usize = 0,
+
+        const Self = @This();
+
+        pub const Component = struct {
+            /// The current component's path name, e.g. 'b'.
+            /// This will never contain path separators.
+            name: []const T,
+            /// The full path up to and including the current component, e.g. '/a/b'
+            /// This will never contain trailing path separators.
+            path: []const T,
+        };
+
+        const InitError = switch (path_type) {
+            .windows => error{BadPathName},
+            else => error{},
+        };
+
+        /// After `init`, `next` will return the first component (no need to call `first`).
+        /// To iterate backwards, call `last` after `init` and then iterate via `previous` calls.
+        /// On Windows, `error.BadPathName` is returned if the `path` has an explicit namespace prefix
+        /// (\\.\, \\?\, or \??\) or if it is a UNC path with more than two path separators at the beginning.
+        pub fn init(path: []const T) InitError!Self {
+            const root_end_index: usize = switch (path_type) {
+                .posix, .uefi => posix: {
+                    // Root on UEFI and POSIX only differs by the path separator
+                    var root_end_index: usize = 0;
+                    while (true) : (root_end_index += 1) {
+                        if (root_end_index >= path.len or !path_type.isSep(T, path[root_end_index])) {
+                            break;
+                        }
+                    }
+                    break :posix root_end_index;
+                },
+                .windows => windows: {
+                    // Namespaces other than the Win32 file namespace are tricky
```

Comment:
I think a shorter counter-example would be more readable, but I dont have too much of an opinion on this.